File: app/src/main/java/com/akari/ppx/xp/hook/code/assist/CommentTimeHook.java
package com.akari.ppx.xp.hook.code.assist;

import com.akari.ppx.common.utils.Utils;
import com.akari.ppx.common.utils.XSP;
import com.akari.ppx.xp.hook.code.SuperbHook;

import de.robv.android.xposed.XC_MethodReplacement;

import static com.akari.ppx.common.constant.Prefs.EXACT_COMMENT_TIME_FORMAT;
import static com.akari.ppx.common.constant.Prefs.SHOW_COMMENT_TIME;
import static com.akari.ppx.common.constant.Prefs.TODAY_COMMENT_TIME_FORMAT;

public class CommentTimeHook extends SuperbHook {
	@Override
	protected void onHook(ClassLoader cl) {
		final String todayFormat = XSP.gets(TODAY_COMMENT_TIME_FORMAT, "今天HH点mm分ss秒"), exactFormat = XSP.gets(EXACT_COMMENT_TIME_FORMAT, "yyyy-MM-dd HH:mm:ss");
		hookMethod(SHOW_COMMENT_TIME, "com.sup.superb.m_feedui_common.util.a", "a", long.class, "kotlin.jvm.functions.Function0", new XC_MethodReplacement() {
			@Override
			protected Object replaceHookedMethod(MethodHookParam param) throws Throwable {
				return Utils.ts2date((Long) param.args[0], Utils.isToday((Long) param.args[0]) ? todayFormat : exactFormat, false);
			}
		});

	}
}

File: app/src/main/java/com/akari/ppx/xp/hook/assist/CommentTimeHook.kt
@file:Suppress("unused")

package com.akari.ppx.xp.hook.assist

import com.akari.ppx.data.XPrefs
import com.akari.ppx.utils.getDiffDays
import com.akari.ppx.utils.replaceMethod
import com.akari.ppx.utils.splitByOr
import com.akari.ppx.utils.ts2Date
import com.akari.ppx.xp.Init.inexactDate
import com.akari.ppx.xp.Init.inexactDateClass
import com.akari.ppx.xp.hook.SwitchHook

class CommentTimeHook : SwitchHook("show_exact_comment_time") {
    override fun onHook() {
        val recentFormat = XPrefs<String>("recent_time_format").splitByOr()
        val exactFormat = XPrefs<String>("exact_time_format")
        inexactDateClass!!.replaceMethod(
            inexactDate(),
            Long::class.java,
            "kotlin.jvm.functions.Function0"
        ) { param ->
            val ts = param.args[0] as Long
            runCatching {
                ts.ts2Date(ts.getDiffDays().let { recentFormat[it] })
            }.getOrElse {
                ts.ts2Date(exactFormat)
            }
        }
    }
}


